package foo.bar.test.everything;


message AllVarintTypes {
  optional int32 a_signed_int32 = 1;
  optional sint32 a_signed_zigzag_int32 = 2;
  optional uint32 an_unsigned_int32 = 3;
  optional int64 a_signed_int64 = 11;
  optional sint64 a_signed_zigzag_int64 = 12;
  optional uint64 an_unsigned_int64 = 13;
  optional bool a_boolean = 20;
}

message AllFixedInt32Types {
  optional fixed32 an_unsigned_int32 = 1;
  optional sfixed32 a_signed_int32 = 2;
  optional float a_float = 3;
}

message AllFixedInt64Types {
  optional fixed64 an_unsigned_int64 = 1;
  optional sfixed64 a_signed_int64 = 2;
  optional double a_double = 3;
}

message AllLenFieldTypes {
  optional string a_string = 1;
  optional bytes some_bytes = 2;
}

message NestedLevel3 {
  optional string a_string = 1;
}

message NestedLevel2 {
  optional NestedLevel3 level3 = 1;
}

message NestedLevel1 {
  optional NestedLevel2 level2 = 1;
}

message ComplexNestedMessage {
  optional AllFixedInt32Types all_fixed_32 = 1;
  optional AllFixedInt64Types all_fixed_64 = 2;
  optional AllVarintTypes all_varint = 3;
  optional AllLenFieldTypes all_len = 4;
  optional NestedLevel1 level1 = 5;

}